### PR TITLE
Add ECS and Logs attributes based on Metadata v4 endpoint

### PR DESCRIPTION
**Pull request by mmanciop**
## Changes

Implement `aws.ecs` and `aws.log` resource attributes with data from the ECS Metadata endpoint v4.

* [X] Appropriate `CHANGELOG.md` updated for non-trivial changes

### CODE REVIEW

**Comment by utpilla** [comment]
@mmanciop Please sign the CLA.

File: src/OpenTelemetry.Contrib.Extensions.AWSXRay/CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,8 @@
 * Enhancement - AWSXRayIdGenerator - Generate X-Ray IDs with global Random
   instance instead of recreating with ThreadLocal
   ([#380](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/380))
+* Enhancement - AWSECSResourceDetector - Implement `aws.{ecs.*,log.*}` resource
+  attributes with data from ECS Metadata endpoint v4
```
Comment (Kielek) [inline src/OpenTelemetry.Contrib.Extensions.AWSXRay/CHANGELOG.md:9]:
```suggestion
* Enhancement - AWSECSResourceDetector - Implement `aws.{ecs.*,log.*}` resource
  attributes with data from ECS Metadata endpoint v4
  ([#875](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/875))
```

File: src/OpenTelemetry.Contrib.Extensions.AWSXRay/OpenTelemetry.Contrib.Extensions.AWSXRay.csproj [hunk 2]
```diff
@@ -1,13 +1,13 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks>net452;netstandard2.0</TargetFrameworks>
+    <TargetFrameworks>net461;netstandard2.0</TargetFrameworks>
```
Comment (Kielek) [inline src/OpenTelemetry.Contrib.Extensions.AWSXRay/OpenTelemetry.Contrib.Extensions.AWSXRay.csproj:3]: If we decide to change minimal .NET Framermwprk it should be raised to .NET Framework 4.6.2. Any older version is not supported by MS. 

This change should be included in CHANGELOG.

File: src/OpenTelemetry.Contrib.Extensions.AWSXRay/OpenTelemetry.Contrib.Extensions.AWSXRay.csproj [hunk 3]
```diff
@@ -1,13 +1,13 @@
...
     <Description>OpenTelemetry extensions for AWS X-Ray.</Description>
     <MinVerTagPrefix>Extensions.AWSXRay-</MinVerTagPrefix>
   </PropertyGroup>
 
   <ItemGroup>
     <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
-    <PackageReference Include="OpenTelemetry" Version="1.1.0" />
+    <PackageReference Include="OpenTelemetry" Version="$(OpenTelemetryCoreLatestVersion)" />
```
Comment (Kielek) [inline src/OpenTelemetry.Contrib.Extensions.AWSXRay/OpenTelemetry.Contrib.Extensions.AWSXRay.csproj:10]: This change should be included in CHANGELOG.

File: src/OpenTelemetry.Contrib.Extensions.AWSXRay/OpenTelemetry.Contrib.Extensions.AWSXRay.csproj (hunk 2 above)
Comment (mmanciop) [reply src/OpenTelemetry.Contrib.Extensions.AWSXRay/OpenTelemetry.Contrib.Extensions.AWSXRay.csproj:3]: I am not convinced we need to update, this change is likely something we can revert. The reason why the PR is in draft, besides the missing tests, is that I am struggling big times with my dev setup atm, and this change is because of it ;-)

Comment (mmanciop) [reply src/OpenTelemetry.Contrib.Extensions.AWSXRay/OpenTelemetry.Contrib.Extensions.AWSXRay.csproj]: Actually, it's probably better to raise to `net462`. Done in f90fce5.

File: src/OpenTelemetry.Contrib.Extensions.AWSXRay/OpenTelemetry.Contrib.Extensions.AWSXRay.csproj (hunk 3 above)
Comment (mmanciop) [reply src/OpenTelemetry.Contrib.Extensions.AWSXRay/OpenTelemetry.Contrib.Extensions.AWSXRay.csproj:10]: Done in f90fce5.

File: src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs [hunk 4]
```diff
@@ -34,18 +38,30 @@ public class AWSECSResourceDetector : IResourceDetector
     /// <returns>List of key-value pairs of resource attributes.</returns>
     public IEnumerable<KeyValuePair<string, object>> Detect()
```
Comment (Kielek) [inline src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs:39]: There is some changes in public API in this package.
I think it will be great to change the implementation from `OpenTelemetry.Contrib.Extensions.AWSXRay.Resources.IResourceDetector` to `OpenTelemetry.Resources.IResourceDetector` and remove the old one interface.

It is candidate do separate PR.

File: src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs [hunk 5]
```diff
@@ -34,18 +38,30 @@ public class AWSECSResourceDetector : IResourceDetector
...
     {
-        List<KeyValuePair<string, object>> resourceAttributes = null;
-
         if (!this.IsECSProcess())
         {
-            return resourceAttributes;
+            return new List<KeyValuePair<string, object>>();
```
Comment (Kielek) [inline src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs:43]: Changed behavior from returning null to empty list. Is it expected?

**Comment by mmanciop** [comment]
I am actually not 100% sure yet of the patchset wrt the implementation of the IResourceDetector interface. I am going to try it out on a test app.

File: src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs (hunk 4 above)
Comment (mmanciop) [reply src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs:39]: That is what I am doing now, but it is across the entire .csproj

Comment (mmanciop) [reply src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs:39]: That is what I am doing now, and it cannot be a follow-up PR (unless I am mistaken) if we want to follow the train of the last major sdk release. Unfortunately, it is across the entire .csproj

File: src/OpenTelemetry.Contrib.Extensions.AWSXRay/.publicApi/net462/PublicAPI.Shipped.txt [hunk 6]
```diff
@@ -11,11 +11,9 @@ OpenTelemetry.Contrib.Extensions.AWSXRay.Resources.ResourceDetectorUtils
 OpenTelemetry.Contrib.Extensions.AWSXRay.Resources.ResourceDetectorUtils.ResourceDetectorUtils() -> void
 OpenTelemetry.Contrib.Extensions.AWSXRay.Trace.AWSXRayPropagator
 OpenTelemetry.Contrib.Extensions.AWSXRay.Trace.AWSXRayPropagator.AWSXRayPropagator() -> void
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/OpenTelemetry.Contrib.Extensions.AWSXRay/.publicApi/net462/PublicAPI.Shipped.txt:13]: Done in 73ff05c

File: src/OpenTelemetry.Contrib.Extensions.AWSXRay/.publicApi/netstandard2.0/PublicAPI.Shipped.txt [hunk 7]
```diff
@@ -20,11 +20,9 @@ OpenTelemetry.Contrib.Extensions.AWSXRay.Resources.ResourceDetectorUtils
 OpenTelemetry.Contrib.Extensions.AWSXRay.Resources.ResourceDetectorUtils.ResourceDetectorUtils() -> void
 OpenTelemetry.Contrib.Extensions.AWSXRay.Trace.AWSXRayPropagator
 OpenTelemetry.Contrib.Extensions.AWSXRay.Trace.AWSXRayPropagator.AWSXRayPropagator() -> void
-OpenTelemetry.Resources.ResourceBuilderExtensions
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/OpenTelemetry.Contrib.Extensions.AWSXRay/.publicApi/netstandard2.0/PublicAPI.Shipped.txt:23]: If I understood right, done in 8a73d34

File: src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs (hunk 5 above)
Comment (mmanciop) [reply src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs:43]: It is fine according to the API, and null-safe is usually better that nullable, but it is not a necessary change by any means

File: src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs (hunk 4 above)
Comment (mmanciop) [reply src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs]: Work started in c213eadc2bec3a67c433ccc83a7584d178d98001

File: src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs (hunk 5 above)
Comment (mmanciop) [reply src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs]: I reverted the change in behavior and align it with the other detectors. I like more detectors returning an empty resource than a `null` one for nullability, but is not something I would cling to in a PR like this.

**Comment by Kielek** [comment]
@srprash, @lupengamzn could you please check?

File: src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSEC2ResourceDetector.cs [hunk 8]
```diff
@@ -52,7 +54,7 @@ public class AWSEC2ResourceDetector : IResourceDetector
             AWSXRayEventSource.Log.ResourceAttributesExtractException(nameof(AWSEC2ResourceDetector), ex);
         }
 
-        return resourceAttributes;
+        return resourceAttributes != null ? new Resource(resourceAttributes) : null;
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSEC2ResourceDetector.cs]: Done in 3155f3d

File: src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSEBSResourceDetector.cs [hunk 9]
```diff
@@ -64,7 +66,7 @@ public class AWSEBSResourceDetector : IResourceDetector
             AWSXRayEventSource.Log.ResourceAttributesExtractException(nameof(AWSEBSResourceDetector), ex);
         }
 
-        return resourceAttributes;
+        return resourceAttributes != null ? new Resource(resourceAttributes) : null;
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSEBSResourceDetector.cs]: Done in 3155f3d

File: src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs [hunk 10]
```diff
@@ -31,42 +37,140 @@ public class AWSECSResourceDetector : IResourceDetector
...
-    public IEnumerable<KeyValuePair<string, object>> Detect()
+    /// <returns>Resource.</returns>
+    public Resource Detect()
     {
-        List<KeyValuePair<string, object>> resourceAttributes = null;
-
         if (!this.IsECSProcess())
         {
-            return resourceAttributes;
+            return null;
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs]: Done in 3155f3d

File: src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs [hunk 11]
```diff
@@ -31,42 +37,140 @@ public class AWSECSResourceDetector : IResourceDetector
...
 
+    internal static List<KeyValuePair<string, object>> ExtractMetadataV4ResourceAttributes()
+    {
+        var metadataV4Url = Environment.GetEnvironmentVariable(AWSECSMetadataURLV4Key);
+        if (metadataV4Url == null)
+        {
+            return new List<KeyValuePair<string, object>>();
+        }
+
+        var httpClientHandler = new HttpClientHandler();
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs]: Yes it is, done in 019d1bc

File: src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs [hunk 12]
```diff
@@ -31,42 +37,140 @@ public class AWSECSResourceDetector : IResourceDetector
...
+        var launchType = GetStringOrThrow(taskResponse, "LaunchType") switch
+        {
+            string type when "ec2".Equals(type.ToLower()) => AWSSemanticConventions.ValueEcsLaunchTypeEc2,
+            string type when "fargate".Equals(type.ToLower()) => AWSSemanticConventions.ValueEcsLaunchTypeFargate,
+            _ => null,
+        };
+
+        if (launchType == null)
+        {
+            throw new ArgumentOutOfRangeException($"Unrecognized launch type '{taskResponse["LaunchType"]}'");
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs:130]: I think that, from the PoV of semantic conventions, it would be exceptional:

```
aws.ecs.launchtype MUST be one of the following:

Value | Description
-- | --
ec2 | ec2
fargate | fargate
```

From https://opentelemetry.io/docs/reference/specification/resource/semantic_conventions/cloud_provider/aws/ecs/ .

Now, we might argue "But what if the Semantic Conventions are incomplete?" Today, I don't think they are. There is, technically, a _third_ launch type: `external`: https://docs.aws.amazon.com/AmazonECS/latest/developerguide/launch_types.html. But, to the best of my knowledge, there is no Metadata endpoint v4 on external VMs in ECS, so it would not apply for this detector.

File: src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs [hunk 13]
```diff
@@ -31,42 +37,140 @@ public class AWSECSResourceDetector : IResourceDetector
...
-        return resourceAttributes;
+        return new Resource(resourceAttributes);
+    }
+
+    internal static string GetStringOrThrow(JObject obj, string fieldName)
+    {
+        var value = obj[fieldName]?.ToString();
+        if (value == null)
+        {
+            throw new ArgumentNullException($"The expected '{fieldName}' field is missing");
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs:81]: I think you have a point about the access to the `LogDriver` field: if no logging is required, as far as I can tell, it will be unspecified. The other fields accessed via `GetStringOrThrow` are part of the ECS Metadata v4 contract as I understand it.

[the remark this answers is not in the document]
Comment (mmanciop) [reply src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs]: Done in f35d9eb

File: src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/AWSECSResourceDetector.cs]: Just in case I am wrong, I added some resilience in f35d9eb

**Review by atshaw43 — APPROVED** [review]
The code looks good to me.

Could this change break existing customers by changing the returned data type from "IEnumerable<KeyValuePair<string, object>>" to "Resource"?

Seems this will need a major version change?

**Comment by mmanciop** [comment]
Given the fact that we also increased the Otel baseline, yes, a major may be in order.

**Comment by cijothomas** [comment]
> The code looks good to me.
> 
> Could this change break existing customers by changing the returned data type from "IEnumerable<KeyValuePair<string, object>>" to "Resource"?
> 
> Seems this will need a major version change?

Agree. If the package was already stable, then this breaking change would need major version bump. If it was pre-release, a good callout in the changelog would be sufficient.

**Comment by cijothomas** [comment]
> Given the fact that we also increased the Otel baseline, yes, a major may be in order.

Bumping Otel version should not require major version bump. It follows semver, and is backward compatible.

**Comment by cijothomas** [comment]
> @srprash, @lupengamzn could you please check?

Would you review? And also comment on the breaking change/major version bump as well?

**Comment by Kielek** [comment]
> > The code looks good to me.
> > Could this change break existing customers by changing the returned data type from "IEnumerable<KeyValuePair<string, object>>" to "Resource"?
> > Seems this will need a major version change?
> 
> Agree. If the package was already stable, then this breaking change would need major version bump. If it was pre-release, a good callout in the changelog would be sufficient.

Just a reminder, in case of major bump, the whole package should be renamed. At least `.Contrib` infix has to be removed.

**Comment by mmanciop** [comment]
> > > The code looks good to me.
> > > Could this change break existing customers by changing the returned data type from "IEnumerable<KeyValuePair<string, object>>" to "Resource"?
> > > Seems this will need a major version change?
> > 
> > 
> > Agree. If the package was already stable, then this breaking change would need major version bump. If it was pre-release, a good callout in the changelog would be sufficient.
> 
> Just a reminder, in case of major bump, the whole package should be renamed. At least `.Contrib` infix has to be removed.

If renaming the package is being considered, could we please replace `XRay` with `AWS` in the name? Not all contributors are working for AWS *cough* *cough*, and the package is useful beyond XRay ([case in point](https://docs.lumigo.io/docs/supported-semantic-conventions)).

**Comment by Kielek** [comment]
@srprash promised to review within next 2 werks.

**Comment by Kielek** [comment]
@srprash?

**Review by srprash — APPROVED** [review]

**Comment by srprash** [comment]
>>@srprash, @lupengamzn could you please check?
>
> Would you review? And also comment on the breaking change/major version bump as well?

I agree that this must go out as a major version bump. 

> Just a reminder, in case of major bump, the whole package should be renamed. At least .Contrib infix has to be removed.

Yup. We can do that to be inline with the rest of the packages in the repo. I wish we can do the same for the [OpenTelemetry.Contrib.Instrumentation.AWS](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Contrib.Instrumentation.AWS) package as well while bumping its major version.

> If renaming the package is being considered, could we please replace XRay with AWS in the name?

I have no issues in dropping XRay and renaming the package to be `OpenTelemetry.Contrib.Extensions.AWS`. We had it with XRay because initially the package contained only the id-generator and the propagator very specific to AWS XRay.

**Comment by Kielek** [comment]
@mmanciop, finally we have an approval from the package owner.
Could you please solve the conflicts? Then we should be able to make final review and merge.

**Comment by mmanciop** [comment]
@Kielek @srprash I solved the conflicts and did some cleanup, but I am assuming that version bumping, extension renaming and the rest will be performed by the package owners, right?

**Comment by mmanciop** [comment]
> @mmanciop, finally we have an approval from the package owner. Could you please solve the conflicts? Then we should be able to make final review and merge.

Now that I found out why `PublicApiAnalyzers` did not run in my setup (#1071), I dearly hope this commit is it :-)

**Review by Kielek — APPROVED** [review]
